File: bin/day-7.dart
/**
 * This is a mostly direct port of the python version. its still about 5x slower
 * and I'm not entirely sure why.
 */

import 'dart:io';

Map<String, Function> ops = {
  'NOT': (int v) => ~v,
  'AND': (int a, int b) => a & b,
  'OR': (int a, int b) => a | b,
  'RSHIFT': (int a, int b) => a >> b,
  'LSHIFT': (int a, int b) => a << b
};


main() {
  Map<String, dynamic> wires = {};
  List<String> lines = new File("./inputs/day-7.txt").readAsLinesSync();
  Stopwatch watch = new Stopwatch();


  for (String line in lines) {
    List<String> parts = line.split(" -> ");
    wires[parts[1]] = parts[0];
  }
  watch.start();
  int evaluate(key) {
    if (key is num) {
      return key;
    } else if (int.parse(key, onError: (e) => null) != null) {
      return int.parse(key);
    } else {
      var val = wires[key];
      if (val is num) {
        return val;
      } else if (int.parse(val, onError: (e) => null) != null) {
        wires[key] = int.parse(val);
      } else if (wires.containsKey(val)) {
        wires[key] = evaluate(val);
      } else {
        List<String> parts = val.split(' ');
        if (parts.length == 3) {
          // binary
          wires[key] = ops[parts[1]](evaluate(parts[0]), evaluate(parts[2]));
        } else if (parts.length == 2) {
          wires[key] = ops[parts[0]](evaluate(parts[1]));
        }
      }
      return wires[key];
    }
  }

  int answer = evaluate('a');
  watch.stop();
  print("a = $answer (${watch.elapsedMilliseconds}ms)");

}

File: bin/day-7.py
import time

ops = {
    'OR': lambda a, b: a | b,
    'AND': lambda a, b: a & b,
    'NOT': lambda a: ~a,
    'RSHIFT': lambda a, b: a >> b,
    'LSHIFT': lambda a, b: a << b
}


def evaluate(wire):

    # 'wire' might be a literal value, because we optimistically call
    # evaluate on each operand in the complex expressions
    if isinstance(wire, int):
        return wire
    elif wire.isdigit():
        return int(wire)

    # it's an actual wire, so we're going to look it up in the dict
    # and evaluate whatever we find there.  then we re-assign it to the wire
    # label for other dependent wires to re-use the value without re-evaluating
    # the whole thing
    val = wires[wire]
    if isinstance(val, int):
        # it's an actual number
        return val
    elif val.isdigit():
        # it's a stringified number
        wires[wire] = int(val)
    elif val in wires:
        # it's a wire label
        wires[wire] = evaluate(val)
    else:
        # it's a complex expression, parse and evaluate it's parts
        parts = val.split(' ')
        if len(parts) == 3:
            # a binary operator
            wires[wire] = ops[parts[1]](evaluate(parts[0]), evaluate(parts[2]))
        elif len(parts) == 2:
            # a unary opeartor
            wires[wire] = ops[parts[0]](evaluate(parts[1]))

    return wires[wire]

wires = {}

now = time.time()

for line in open("../inputs/day-7.txt"):
    l, r = line.split(" -> ")
    wires[r.strip()] = l

print evaluate('a'),

print 'in', int((time.time() - now) * 1000000), 'microseconds'